---
title: "BioInfomatics Report - Assignment 3" 
author: "Greg Farmer"
date: '2022-05-22'
output: html_document
---

### Intro 
Source code: https://github.com/GregFarm/BioInformatics-Report-3-/blob/main/BioInfo%20Report.Rmd 

Before knitting the Rmarkdown, please ensure that the correct libraries are loading in. The libraries needed are seqinr, kableExtra, R.utils. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("seqinr")
library("kableExtra")
library("R.utils")
library("dplyr")
```

# Assignement Part 1 - Gene Expression 
### Questions 1 - Read in the gene_expression.cvs file and show the values of the first row 
```{r, mychuck1.1}
URL = "https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/gene_expression.tsv"
gene_exp <- read.table(URL, stringsAsFactors = FALSE, header = TRUE)
gene_exp[1:6,]
```


### Questions 2 - Make a new column which is the mean of the other columns. Show a table of values for the first six genes


```{r, mychunk1.2}
gene_mean <- (gene_exp$GTEX.1117F.0226.SM.5GZZ7 + gene_exp$GTEX.1117F.0426.SM.5EGHI + gene_exp$GTEX.1117F.0526.SM.5EGHJ)/3
gene_exp['Mean'] <- signif(gene_mean, digits = 6)
gene_exp[1:6,]

```
### Questions 3 - List the 10 genes with the highest mean expression
```{r, mychunk1.3}
high_mean <- gene_exp[order(-gene_exp$Mean),]
high_mean[1:10,]
```

### Questions 4 - Determine the number of genes with a mean <10
```{r, mychunk1.4}
true_sum <- gene_exp$Mean < 10
sum(true_sum, na.rm = TRUE)
```
If the mean was higher then 10 it converted it into a TRUE value, than the total amount of TRUE values was summed together. 
It was found that 35,988 mean gene expression was higher then 10. 

### Questions 5 - Make a histogram plot of the mean values and include it into your report.
```{r, mychunk1.5}
x_vals <- gene_exp$Mean
h <-hist(x_vals,main="Mean Gene Expression Values", xlab="Mean Gene Expression",breaks= "scott" ,xlim=c(0,10000),ylim = c(0,50000), col = "blue")
 
```
### Questions 6 - Import this csv file into an R object. What are the column names?
```{r, mychunk1.6}
URL2 <- "https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/growth_data.csv"
grow_data <- read.csv(URL2)
colnames(grow_data)
```
Simply using the `colnames` function to return the column names, being:
Site
TreeID
Circumf_2005_cm
Circumf_2010_cm
Circumf_2015_cm
Circumf_2020_cm

### Questions 7 - Calculate the mean and standard deviation of tree circumference at the start and end of the study at both sites.
```{r, mychunk1.7}
#northeast data
NE_data <-grow_data[grep("northeast", grow_data$Site),]
NE_mean05 <- mean(NE_data$Circumf_2005_cm, na.rm = TRUE)
NE_SD05 <- sd(NE_data$Circumf_2005_cm, na.rm = TRUE)
NE_mean20 <- mean(NE_data$Circumf_2020_cm, na.rm = TRUE)
NE_SD20 <- sd(NE_data$Circumf_2020_cm, na.rm = TRUE)
cat(paste("NorthEast 2005 mean =", NE_mean05), sep="\n")
cat(paste("NorthEast 2005 standard deviation =", signif(NE_SD05,digits = 3)), sep="\n")
cat(paste("NorthEast 2020 mean =", NE_mean20), sep="\n")
cat(paste("NorthEast 2020 standard deviation =", signif(NE_SD20,digits = 3)), sep="\n")

#southwest data
SW_data <-grow_data[grep("southwest", grow_data$Site),]
SW_mean05 <- mean(SW_data$Circumf_2005_cm, na.rm = TRUE)
SW_SD05 <- sd(SW_data$Circumf_2005_cm, na.rm = TRUE)
SW_mean20 <- mean(SW_data$Circumf_2020_cm, na.rm = TRUE)
SW_SD20 <- sd(SW_data$Circumf_2020_cm, na.rm = TRUE)
cat(paste("SouthWest 2005 mean =", SW_mean05), sep="\n")
cat(paste("SouthWest 2005 standard deviation =", signif(SW_SD05,digits = 3)), sep="\n")
cat(paste("SouthWest 2020 mean =", SW_mean20), sep="\n")    
cat(paste("SouthWest 2020 standard deviation =", signif(SW_SD20,digits = 3)), sep="\n")         
```
This returns all the values individually. 

### Questions 8 - Make a box plot of tree circumference at the start and end of the study at both sites.
```{r, mychunk1.8}
boxplot(NE_data$Circumf_2005_cm, NE_data$Circumf_2020_cm,
        SW_data$Circumf_2005_cm, SW_data$Circumf_2020_cm,
        main ="Plant Circumference Data at Different Sites Over Time",
        names = c("Northeast 2005","Northeast 2020","Southwest 2005","Southwest 2020"),
        ylab = "Mean Tree Circumference (cm)",
        col = "green"
        )
```
### Questions 9 - Calculate the mean growth over the last 10 years at each site.
```{r, mychunk1.9}
NE_tenyear <-NE_data[,c("Circumf_2010_cm", "Circumf_2015_cm", "Circumf_2020_cm")]
NE_tenyear_mean <- colMeans(NE_tenyear, na.rm = TRUE)
sum(NE_tenyear_mean)/3

SW_tenyear <-SW_data[,c("Circumf_2010_cm", "Circumf_2015_cm", "Circumf_2020_cm")]
SW_tenyear_mean <- colMeans(SW_tenyear, na.rm = TRUE)
sum(SW_tenyear_mean)/3
```
Above is the mean growth for each site over the last 10 years. 
The mean growth over a ten year period, 2010 to 2020, was 30.0 cm at the north east location and 25.7 cm at the south west location. 

### Questions 10 - Use the t.test and wilcox.test functions to estimate the p-value that the 10 year growth is different at the two sites.
```{r, mychunk 1.10}
ttest <- t.test(NE_tenyear, SW_tenyear)
ttest$p.value
```



# Assignement Part 2 - Examining biological sequence diversity

For this section the organisms being used are E.coli (SPECIFY) and Mycobacterium tuberculosis (GCA_001318445)

### Questions 1.1 - Download the whole set of coding DNA sequences for E. coli and your organism of interest. 
```{r, mychunk2.1}
URL_M <- "http://ftp.ensemblgenomes.org/pub/bacteria/release-53/fasta/bacteria_4_collection/mycobacterium_tuberculosis_gca_001318445/cds/Mycobacterium_tuberculosis_gca_001318445.6505_5_10.cds.all.fa.gz"
download.file(URL_M, destfile = "Mturb_cds.fa.gz")
URL_E="http://ftp.ensemblgenomes.org/pub/bacteria/release-53/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cds/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cds.all.fa.gz"
download.file(URL_E, destfile="ecoli_cds.fa.gz")

gunzip("Mturb_cds.fa.gz")
gunzip("ecoli_cds.fa.gz")

ecoli_cds <- seqinr::read.fasta("ecoli_cds.fa")
mturb_cds <- seqinr::read.fasta("Mturb_cds.fa")

```
### Question 1.2 - How many coding sequences are present in these organisms? How much coding DNA is there in total for these two organisms? Describe any differences between the two organisms.

```{r, mychunk2.2}
length(mturb_cds)
length(ecoli_cds)
```
The total amount of coding sequences in M. tuberculosis is 4,752 and for E. coli the total amount of coding sequences is 4239.

```{r, mychunk2.3}
m_len<-sapply(X = mturb_cds, FUN = length)
sum(m_len)
e_len<-sapply(X = ecoli_cds, FUN = length)
sum(e_len)
```
M. tuberculosis has a total of 3,072,513 coding nucleotides.
E. coli has a total of 3,978,528 coding nucleotides. 

M. tuberculosis has 513 more coding sequences, yet nearly 1 million less coding nucleotides comapred to E. coli. 

### Question 2.1 - Calculate the length of all coding sequences in these two organisms. Make a boxplot of coding sequence length in these organisms.

```{r, mychunk2.4}
boxplot(m_len, e_len, names = c("M. tuberculosis","E. coli"),
        main = "Coding Sequence Comaprison: M. tuberculosis and E. coli",
          ylab = "Base Pair Length (bp)",
        col = c("green","red"))

```

### Question 2.2 - What is the mean and median coding sequence length of these two organisms? Describe any differences between the two organisms

```{r, mychunk2.5}
signif(mean(m_len), digits = 4) 
median(m_len)

signif(mean(e_len), digits = 4) 
median(e_len)
```
The mean and median coding sequences length for M. tuberculosis was shown to be 647bp and 501bp respectively. 
The mean and median coding sequences length for E. coli was shown to be 939 bp and 801bp respectively. (MORE DETAILS)

### Question 3.1 - Calculate the frequency of DNA bases in the total coding sequences for both organisms. Perform the same calculation for the total protein sequence.   
```{r, muchunk2.6}
# nt count 
m_dna <- unlist(mturb_cds)
m_dna_comp <-count(m_dna, wordsize = 1)
m_dna_comp

e_dna <- unlist(ecoli_cds)
e_dna_comp <-count(e_dna, wordsize = 1)
e_dna_comp

```
```{r, mychunk2.8}
#AA count 
#MINUS THE TERMINATION CODON

m_prot <- lapply(mturb_cds, translate)
aa <- unique(m_prot[[2]])
aa <- aa[aa != "*"]
m_comp <- unlist(m_prot)
m_aa <- count(m_comp, wordsize = 1 , alphabet = aa)


e_prot <- lapply(X=ecoli_cds, FUN = translate)
e_comp<- unlist(e_prot)
e_aa <-count(e_comp, wordsize = 1, alphabet = aa)
```

### Question 3.2 -Create bar plots for nucleotide and amino acid frequency. Describe any differences between the two organisms.

```{r, mychunk2.7}
#NT barplot
barplot(as.matrix(rbind(m_dna_comp, e_dna_comp)), beside = TRUE, 
        names = c("a","c","g","t"), col = c("red","blue"),
        ylim = c(0,1200000),
        xlab = "Nucleotide", ylab = "Nucleotide Frequency (bp)",
        main = "Nucleotide Fequencies Comparison Between M. tuberculosis & E. coli")
legend("topleft", c("M.tuberculosis", "E. coli"), horiz = T, col = c("red", "blue"), fill = c("red", "blue"))
```


```{r, mychunk2.9}
#AA bar plot
barplot(as.matrix(rbind(m_aa, e_aa)), beside = TRUE, ylim = c(0,160000), 
        col = c("red","blue"),
        xlab = "Amino Acids", ylab = "Amino Acid Frequency",
        main = "Amino Acid Fequencies Comparison Between M. tuberculosis & E. coli")
legend("topright", c("M.tuberculosis", "E. coli"), horiz = T, col = c("red", "blue"), fill = c("red", "blue"))
```
DESCRIBE ANY DIFFERENCES

### Question 4.1 -Create a codon usage table and quantify the codon usage bias among all coding sequences.

```{r, mychunk2.10}
m_codon <- uco(m_dna, index="rscu", as.data.frame = TRUE)
m_codon%>%
  kbl(caption = "M. tuberculosis Codon Usage")%>%
  kable_styling(bootstrap_option = c("striped", "hover", "responsive", "condensed"), full_width = TRUE, fixed_thead=TRUE )

e_codon <-uco(e_dna, index="rscu", as.data.frame = TRUE)
e_codon[order(e_codon$RSCU, decreasing = TRUE),]%>%
  kbl(caption = "E coli Codon Usage")%>%
  kable_styling(bootstrap_option = c("striped", "hover", "responsive", "condensed"), full_width = TRUE, fixed_thead=TRUE )

```
DESCRIBE BIAS


### Question 5.1 - In the organism of interest, identify 10 protein sequence k-mers of length 3-5 which are the most overand under-represented k-mers in your organism of interest.

```{r, mychunk2.11}
options(scipen=999)
#M. tuberculosis k-mer (3,4,5)
m_freq3 <-  count(m_comp,wordsize=3,alphabet=aa,freq=TRUE)
m_freq4 <-  count(m_comp,wordsize=4,alphabet=aa,freq=TRUE)
m_freq5 <-  count(m_comp,wordsize=5,alphabet=aa,freq=TRUE)

m_sorted_high <- head(sort(m_freq3, decreasing = TRUE),n=5L)
m_sorted_high
 AAA         LAA         AAL         ALA         VAA 
m_sorted_low <-head(sort(m_freq3, decreasing = FALSE),n=5L)
m_sorted_low
YHM            WKM            WKW            WMN            WNK 


#E.coli k-mer (3,4,5)
e_freq3 <-  count(e_comp,wordsize=3,alphabet=aa,freq=TRUE)
e_freq4 <-  count(e_comp,wordsize=4,alphabet=aa,freq=TRUE)
e_freq5 <-  count(e_comp,wordsize=5,alphabet=aa,freq=TRUE)

View(e_freq3)

e_sorted_high <- e_freq3[c(1,2889,161,153,5777)]
e_sorted_high

e_sorted_low <-head(sort(e_freq3, decreasing = FALSE),n=5L)
e_sorted_low


barplot(as.matrix(rbind(m_sorted_high, e_sorted_high)), beside = TRUE)
barplot(as.matrix(rbind(m_sorted_low, e_sorted_low)), beside = TRUE)

```














